ruby/phase-1.md

```diff
@@ -0,0 +1,575 @@
+
+Here's a diagram to demonstrate the whole setup:
+
+<img src="./ruby_threads.png" width="400" align="center" alt="Temporal" />
+
+### Alternatives considered
+
+We've decided against using the `Client` to initialise the `Worker` (something that Go and Python
+SDKs are doing). It seemed like a wrong abstraction and would require making many attributes public
+on the client to make sure the worker can access them (expanding the public API of a client). Using
+a configuration object solves this and opens up further optimisation opportunities (that will be
+proposed in the 2nd phase).
+
+We can potentially use the reactor pattern via Async gem. This has the benefit of giving extra
+performance, however it would drastically reduce the performance unless activities are written in a
+reactor-friendly way. This can be further added as an optional approach to apps already using Async,
+but doesn't feel like a great default option.
+
+Additionally the SDK users will want to leverage multiple CPUs by spawning multiple worker
+processes. We have agreed that this part should be left out of the Ruby SDK and handled by
+developers themselves. Potentially this functionality can be added as a separate Ruby gem.
+
+
+## Activities
+
+Each activity is going to be represented by its own class. This decision comes with a few benefits:
+
+- There's no need to pollute global scope with activity methods (see "Global Scope" section)
+- There's no need to inject global context for communicating with Temporal. It will be available
+  through a private variable on the class instance (`#activity`)
+- It plays nicely with Ruby's OOP model
+- It allows for future extensibility including 3rd party extensions
+- Activity can be easily tested by injecting a test context into the class instead of the real one
+
+By default we'll use the name of the class as activity name (when communicating with the server).
+This is of course overridable in case the SDK users would like to use their own naming schema. And
+we'll use the method `#execute` as the primary entrypoint for executing user code.
+
+```ruby
+class MyActivity < Temporal::Activity
+  def execute(arg_1, arg_2: nil)
+    loop do
+      sleep 1
+      activity.logger.info('❤️')
+      activity.logger.debug("Attempt #{activity.info.attempts}")
+      activity.heartbeat('details')
+    end
+  rescue Temporal::Error::CancellationRequested
+    activity.logger.info('💀')
+    # Accept cancellation request
+    raise Temporal::Error::ActivityCancelled, 'cancellation details here'
+  end
+end
+```
+
+Notes:
+
+- `activity` is an instance of `Temporal::Activity::Context`
+- `activity.logger` will use a logging adapter configure via `Temporal::Configuration` (more on
+  this in the 2nd phase)
+- Common logging context can be configured by using middleware (more on this in the 2nd phase)
+- `activity.heartbeat` is a synchronous call
+- In case of a cancellation request `Temporal::Error::CancellationRequested` will get raised, which
```

Comment:
Does Ruby support raising an error inside another thread while that thread is running? Or are we gonna raise via heartbeat? (which is somewhat reasonable, though not necessarily how core will send it)

Reply:
We won't be able to raise via heartbeat, what if the activity only heartbeats once and core needs to cancel it?
There should be another way to listen on cancellation in the activity.
In Python we use asyncio task cancellation (for async activities), not sure if there's a standard Ruby cancellation mechanism.
I'd also consider using the same error class that we use to represent failure in the API `CanceledFailure`.

Reply:
Ruby does allow us to raise in a Thread — https://ruby-doc.org/core-2.5.2/Thread.html#method-i-raise. However that's super sketchy as you never know what where will the activity get interrupted.

There is no standard cancellation mechanism in Ruby unfortunately because it inherently doesn't support async programming model (only via additional libraries). Is Core expected to be able to cancel an activity at any time?

@bergundy what does a `CanceledFailure` mean? Is `Failure` synonymous with `Exception`?